project_wip/wip_to_cgs.py

```diff
@@ -0,0 +1,205 @@
+        if not self.project_type_id.wip_account_id:
+            raise ValidationError(
+                _('The project type {} has no WIP (Work In Progress) account defined.')
+                .format(self.project_type_id.name)
+            )
+
+    def _check_project_type_has_cgs_account(self):
+        """Check that the project type has a cgs account defined.
+
+        :raises: ValidationError if no account defined.
+        """
+        if not self.project_type_id.cgs_account_id:
+            raise ValidationError(
+                _('The project type {} has no CGS (Cost of Goods Sold) account defined.')
+                .format(self.project_type_id.name)
+            )
+
+    def _get_posted_unreconciled_wip_lines(self):
+        """Get WIP lines for this project that are posted and unreconciled.
+
+        :rtype: account.move.line recordset
+        """
+        return self.env['account.move.line'].search([
+            ('analytic_account_id', '=', self.analytic_account_id.id),
+            ('account_id', '=', self.project_type_id.wip_account_id.id),
+            ('reconciled', '=', False),
+            ('move_id.state', '=', 'posted'),
+        ])
+
+    def _get_common_wip_to_cgs_move_line_vals(self, wip_line):
+        """Get the account move line vals common to both the debit and the credit part."""
+        return {
+            'name': wip_line.name,
+            'quantity': wip_line.quantity,
+            'product_uom_id': wip_line.product_uom_id.id,
+            'product_id': wip_line.product_id.id,
+            'partner_id': wip_line.partner_id.id,
+            'analytic_account_id': self.analytic_account_id.id,
+            'analytic_tag_ids': [(6, 0, wip_line.analytic_tag_ids.ids)],
+        }
+
+    def _create_wip_to_cgs_account_move(self, wip_line):
+        """Generate an account move to transfer a WIP amount into CGS.
+
+        :param wip_line: the Work In Progress account move line to transfer into CGS.
+        :return: the transfer account move.
+        """
+        wip_reversal_vals = self._get_common_wip_to_cgs_move_line_vals(wip_line)
+        wip_reversal_vals['account_id'] = self.project_type_id.wip_account_id.id
+        wip_reversal_vals['debit'] = wip_line.credit if wip_line.credit else 0
+        wip_reversal_vals['credit'] = wip_line.debit if wip_line.debit else 0
+
+        cgs_vals = self._get_common_wip_to_cgs_move_line_vals(wip_line)
+        cgs_vals['account_id'] = self.project_type_id.cgs_account_id.id
+        cgs_vals['debit'] = wip_line.debit if wip_line.debit else 0
+        cgs_vals['credit'] = wip_line.credit if wip_line.credit else 0
+
+        return self.env['account.move'].create({
+            'journal_id': self.project_type_id.cgs_journal_id.id,
+            'no_analytic_lines': True,
+            'line_ids': [
+                (0, 0, wip_reversal_vals),
+                (0, 0, cgs_vals),
+            ]
+        })
+
+    def _reconcile_wip_move_lines(self, wip_line, wip_reversal_line):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3225679)